src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java

```diff
@@ -382,15 +409,34 @@ void addEvent (Event r) {
     class Dispatcher implements Runnable {
 
         private void handleEvent (Event r) {
+
+            // Stopping marking the state as finished if stop is requested,
+            // termination is in progress and exchange count is 0
+            if (r instanceof StopRequestedEvent) {
+                logger.log(Level.TRACE, "Handling Stop Requested Event");
+
+                // checking if terminating is set to true
+                final boolean terminatingCopy = terminating;
+                assert terminatingCopy;
+
+                if (getExchangeCount() == 0 && reqConnections.isEmpty()) {
+                    finishedLatch.countDown();
+                } else {
+                    logger.log(Level.TRACE, "Some requests are still pending");
+                }
+                return;
+            }
+
             ExchangeImpl t = r.exchange;
             HttpConnection c = t.getConnection();
+
             try {
                 if (r instanceof WriteFinishedEvent) {
 
                     logger.log(Level.TRACE, "Write Finished");
                     int exchanges = endExchange();
```

Comment:
startExchange is never called => endExchange can return negative ?

Reply:
I hope that this is a scenario that cannot happen: to end an exchange you would have to start it first.